Feature/bjs2 78058 minio avatar

tanufa25 opened the pull request.

Commit:
BJS2-78058
Добавление аватара пользователя в minio

Commit:
BJS2-78058
Добавление аватара пользователя в minio

FROLOV-KI commented on src/main/java/school/faang/user_service/exception/config/GlobalExceptionHandler.java:
```diff
@@ -111,12 +116,23 @@ public ResponseEntity<ErrorResponse> handleHttpMessageNotReadable(HttpMessageNot
+        return new ResponseEntity<>(error, HttpStatus.PAYLOAD_TOO_LARGE);
+    }
+
+
```
пустая строчка

FROLOV-KI commented on src/main/java/school/faang/user_service/service/image/ImageResizer.java:
```diff
@@ -0,0 +1,53 @@
+        try {
+            BufferedImage originalImage = ImageIO.read(originalFile.getInputStream());
+
+
```
пустая строчка

FROLOV-KI commented on src/main/java/school/faang/user_service/service/image/ImageResizer.java:
```diff
@@ -0,0 +1,53 @@
+            int width = originalImage.getWidth();
+            int height = originalImage.getHeight();
+
+            double scale = 1.0;
```
мб в константу или конфиг?

tanufa25 replied:
Это, по сути, инициализация, а дальше scale высчитывается по размеру изображения

FROLOV-KI commented on src/main/java/school/faang/user_service/service/image/ImageResizer.java:
```diff
@@ -0,0 +1,53 @@
+            int height = originalImage.getHeight();
+
+            double scale = 1.0;
+            if (width > height && width > maxSideSize) {
```
вынести в boolean переменную с осмысленным названием? для читаемости

FROLOV-KI commented on src/main/java/school/faang/user_service/service/image/ImageResizer.java:
```diff
@@ -0,0 +1,53 @@
+            double scale = 1.0;
+            if (width > height && width > maxSideSize) {
+                scale = (double) maxSideSize / width;
+            } else if (height >= width && height > maxSideSize) {
```
+

FROLOV-KI commented on src/main/java/school/faang/user_service/service/image/ImageResizer.java:
```diff
@@ -0,0 +1,53 @@
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
добавить строчку после метода

FROLOV-KI commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -30,4 +56,56 @@ public List<User> getUsersById(List<Long> usersId) {
+        user.setUserProfilePic(userProfilePic);
+        User savedUser =  userRepository.save(user);
+        return savedUser.getUserProfilePic();
+    }
```
пустая строка после метода

FROLOV-KI commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -30,4 +56,56 @@ public List<User> getUsersById(List<Long> usersId) {
+        s3Service.deleteFile(key);
+        s3Service.deleteFile(keyMini);
+        user.setUserProfilePic(null);
+    }
```
+

FROLOV-KI commented on src/main/java/school/faang/user_service/validation/user/UserValidation.java:
```diff
@@ -0,0 +1,33 @@
+
+@Component
+@Slf4j
+public class UserValidation {
```
Мб FileValidation? Или Avatar, словно бы точно не User

Commit:
BJS2-78058
Загрузка аватаров с PathVariable userId, а не для текущего.
Исправление замечаний.

Fagan89 commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -1,23 +1,34 @@
     private final UserServiceFacade userService;
 
-    @GetMapping("/users/{userId}")
+    @GetMapping("/{userId}")
```
я бы URL Это  метода не трогал бы, мы по нему из других сервисов ходим. там в запросох нету "api/v1/" Лучше для аватарок создать отдельный контроллер !!!!
![image](https://github.com/user-attachments/assets/dbc292c3-c765-45fc-8b80-7a9bdf99e9bc)

tanufa25 replied:
Ого, а почему без api/v1? Может, всё же лучше переделать вызов?
Или в этом есть какой-то скрытый смысл?

Fagan89 commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -1,23 +1,34 @@
 import java.util.List;
 
 @RestController
+@RequestMapping("/api/v1/users")
```
невозможно будет запросить из другого сервиса

Fagan89 commented on src/main/java/school/faang/user_service/dto/resource/S3FileDto.java:
```diff
@@ -0,0 +1,13 @@
+@Data
+@Builder
+public class S3FileDto {
+    private Resource resource;
```
я тут не шарю, но стоит ли передавать в Dto сущность ?

tanufa25 replied:
Это не сущность, это  org.springframework.core.io.Resource, чтобы картинка отображалась

Fagan89 commented on src/main/java/school/faang/user_service/service/image/ImageResizer.java:
```diff
@@ -0,0 +1,58 @@
+
+    public MultipartFile resizeMultipartImage(MultipartFile originalFile, int maxSideSize) {
+        try {
+            BufferedImage originalImage = ImageIO.read(originalFile.getInputStream());
```
создаем InputStream, но не закрываем 

пример: try (InputStream inputStream = file.getInputStream()) {
            s3Client.putObject(request, RequestBody.fromInputStream(inputStream, file.getSize()));

Fagan89 commented on src/main/java/school/faang/user_service/service/image/ImageResizer.java:
```diff
@@ -0,0 +1,58 @@
+                scale = (double) maxSideSize / height;
+            }
+
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```
надо закрыть

Fagan89 commented on src/main/java/school/faang/user_service/service/s3/S3Service.java:
```diff
@@ -0,0 +1,79 @@
+                    .contentType(file.getContentType())
+                    .metadata(Map.of("filename", file.getOriginalFilename()))
+                    .build();
+            client.putObject(request, RequestBody.fromInputStream(file.getInputStream(), file.getSize()));
```
закрыть 

try (InputStream inputStream = file.getInputStream()) {
            s3Client.putObject(request, RequestBody.fromInputStream(inputStream, file.getSize()));
        } catch (IOException e) {

Commit:
BJS2-78058
Minio
Исправление замечаний.

20cognosce commented on src/main/java/school/faang/user_service/dto/resource/S3FileDto.java:
```diff
@@ -0,0 +1,13 @@
+public class S3FileDto {
+    private Resource resource;
+    private String fileName;
+    private String contentType;
```
MediaType enum

20cognosce commented on src/main/java/school/faang/user_service/config/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,32 @@
+    public S3Client amazonS3() {
+        AwsBasicCredentials credentials = AwsBasicCredentials.create(accessKey, secretKey);
+        return S3Client.builder()
+                .region(Region.US_WEST_2)
```
? 
вопрос, не было ли другого
и вопрос, на что это влияет

Если от этого строится часовой пояс, то не гуд

tanufa25 replied:
От региона зависит пинг и стоимость, на часовой пояс не влияет.
Мы с ребятами по s3 когда созванивались, пришли к общему одинаковому подходу по настройкам.

20cognosce commented on src/main/java/school/faang/user_service/config/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,32 @@
+    @Value("${services.s3.endpoint}")
+    private String endpoint;
+
+    @Bean
+    public S3Client amazonS3() {
```
У вас будут конфликты в Колей, у него все те же самые классы созданы

tanufa25 replied:
Не, у Коли другой микросервис

tanufa25 replied:
А вот с Тамирланом могут быть, он генерацию случайных аватаров тоже в userService делает

20cognosce commented on src/test/java/school/faang/user_service/service/user/UserServiceTest.java:
```diff
@@ -6,18 +6,35 @@
+import static school.faang.user_service.util.SettingsConstants.MAX_SIDE_SIZE_MINI;
 
 @ExtendWith(MockitoExtension.class)
 class UserServiceTest {
```
Интеграционный тест на контроллер + S3 сделай, было бы супер
Если есть время и силы конечно

одни юниты это частичное покрытие

20cognosce commented on src/main/java/school/faang/user_service/service/image/InMemoryMultipartFile.java:
```diff
@@ -0,0 +1,62 @@
+import java.io.InputStream;
+
+@RequiredArgsConstructor
+public class InMemoryMultipartFile implements MultipartFile {
```
Готового класса в спринге не было?

tanufa25 replied:
Надо было целую зависимость для него тащить, я решила, что так проще и быстрее

20cognosce commented on src/main/java/school/faang/user_service/service/image/ImageResizer.java:
```diff
@@ -0,0 +1,55 @@
+public class ImageResizer {
+
+    public MultipartFile resizeMultipartImage(MultipartFile originalFile, int maxSideSize) {
+        try (InputStream inputStream = originalFile.getInputStream()) {
+            BufferedImage originalImage = ImageIO.read(inputStream);
+            int width = originalImage.getWidth();
+            int height = originalImage.getHeight();
+            boolean widthAndHeightLessMaxSideSize = width <= maxSideSize && height <= maxSideSize;
+            if (widthAndHeightLessMaxSideSize) {
+                return originalFile;
+            }
+            boolean widthMoreHeight = width > height;
+            double scale = widthMoreHeight
+                    ? (double) maxSideSize / width
+                    : (double) maxSideSize / height;
+
+            try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+                String formatName = getFileExtension(originalFile.getOriginalFilename());
+
+                Thumbnails.of(originalImage)
+                        .scale(scale)
+                        .outputFormat(formatName)
+                        .toOutputStream(outputStream);
```
выглядит очень страшно...
можно не пугать так тех, кто впервые попадет в этот метод?

декомпозировать там на подметоды...

20cognosce commented on src/main/java/school/faang/user_service/service/image/ImageResizer.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private String getFileExtension(String fileName) {
+        if (fileName == null) return "jpg";
+        int dotIndex = fileName.lastIndexOf('.');
+        return (dotIndex >= 0) ? fileName.substring(dotIndex + 1).toLowerCase() : "jpg";
```
Почему так?
Ты работаешь только с картинками

Нужна проверка что расширение есть, и оно в допустимых для картинки

Иначе исключение

20cognosce commented on src/main/java/school/faang/user_service/service/s3/S3Service.java:
```diff
@@ -0,0 +1,83 @@
+    private String bucketName;
+
+    public String uploadFile(String folder, MultipartFile file) {
+        String key = String.format("%s/%d%s", folder, System.currentTimeMillis(), file.getOriginalFilename());
```
это хардкод, вынести паттерн в конфиги

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -30,4 +57,55 @@ public List<User> getUsersById(List<Long> usersId) {
+
+        fileValidation.validateMaxFileSize(file);
+
+        MultipartFile compressedImage = imageResizer.resizeMultipartImage(file, MAX_SIDE_SIZE);
+        String fileKey = s3Service.uploadFile(AVATAR_FOLDER, compressedImage);
+        log.info("Upload avatar, fileKey = {}", fileKey);
+        MultipartFile compressedImageMini = imageResizer.resizeMultipartImage(file, MAX_SIDE_SIZE_MINI);
+        String smallFileKey = s3Service.uploadFile(AVATAR_MINI_FOLDER, compressedImageMini);
+        log.info("Upload avatar, smallFileKey = {}", smallFileKey);
+        UserProfilePic userProfilePic = new UserProfilePic();
+        userProfilePic.setFileId(fileKey);
+        userProfilePic.setSmallFileId(smallFileKey);
+        user.setUserProfilePic(userProfilePic);
+        User savedUser =  userRepository.save(user);
```
расщепляй пространство между строчками

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -30,4 +57,55 @@ public List<User> getUsersById(List<Long> usersId) {
+        String key = user.getUserProfilePic().getFileId();
+        String keyMini = user.getUserProfilePic().getSmallFileId();
+        s3Service.deleteFile(key);
+        s3Service.deleteFile(keyMini);
+        user.setUserProfilePic(null);
```
вот я не совсем согласен

у нас Тамирлан реализует фичу по интеграции с генератором аватарок
Если ты удаляешь аватарку, так надо дефолтную тогда поставить какую-нибудь
И здесь должен быть вызов фичи Тамирлана

20cognosce reviewed (approved):
Нормик, давай к другим темам
